### PR TITLE
Make overlapping_inherent_impls lint a hard error

### LINKED ISSUE #36889: Tracking issue for `overlapping_inherent_impls` compatibility lint
**UPDATE:** [We've decided to convert this warning into a hard error](https://github.com/rust-lang/rust/issues/36889#issuecomment-275166478), but are still waiting for a PR to make that a reality. If you're interested in contributing, [there are directions for writing such a PR available here](https://github.com/rust-lang/rust/issues/36889#issuecomment-287145267).

This tracking issue pertains to a bug fix for https://github.com/rust-lang/rust/issues/22889. The problem was that the compiler was accepting two inherent impls that both defined the same method, which creates an obvious ambiguity:

```rust
struct Foo;

impl Foo {
    fn id() {}
}

impl Foo {
    fn id() {}
}
```

The most obvious fix here is to give one method a distinct name, though sometimes there are other possible rewrites (e.g., one could rewrite the impls to use a double dispatch pattern).

#### Current status

- [x] https://github.com/rust-lang/rust/pull/31925 introduces the `overlapping_inherent_impls` lint as warn-by-default
- [x] https://github.com/rust-lang/rust/pull/36894 makes the `overlapping_inherent_impls` lint deny-by-default
- [ ] PR ? makes the `overlapping_inherent_impls` lint a hard error

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #36889.

### CODE REVIEW

File: src/librustc_typeck/coherence/inherent.rs [hunk 1]
```diff
@@ -295,12 +294,12 @@ impl<'a, 'tcx> InherentOverlapChecker<'a, 'tcx> {
...
-                    let node_id = self.tcx.hir.as_local_node_id(item1).unwrap();
-                    self.tcx.sess.add_lint(lint::builtin::OVERLAPPING_INHERENT_IMPLS,
-                                           node_id,
-                                           self.tcx.span_of_impl(item1).unwrap(),
-                                           msg);
+                    struct_span_err!(self.tcx.sess,
+                                     self.tcx.span_of_impl(item1).unwrap(),
+                                     E0592,
+                                     "duplicate definitions with name `{}`",
+                                     name)
```
Comment (estebank) [inline src/librustc_typeck/coherence/inherent.rs]: The code looks good, but could you add a `span_label` to this diagnostic pointing at both `item1`'s and another pointing at `item2`'s span?

That way the code output would look this way:

```rust
error: duplicate definitions with name `id`
 --> <anon>:4:5
  |
4 |     fn id() {} //~ ERROR duplicate definitions
  |     ^^^^^^^^^^ duplicate definitions for `id`
...
8 |     fn id() {}
  |     ---------- other definition for `id`
```

Comment (topecongiro) [inline src/librustc_typeck/coherence/inherent.rs]: @estebank Thank you for your comment. I will rebase accordingly 😸

**Comment by alexcrichton** [comment]
@bors: retry

* openssl build failure https://github.com/rust-lang/rust/issues/40417

**Comment by alexcrichton** [comment]
@bors: retry

* ar.exe error https://github.com/rust-lang/rust/issues/40546

**Comment by estebank** [comment]
@topecongiro, is it me or the rebased version that was merged didn't have the "hard error" code? If that is the case, could you please send a PR with it?

**Comment by topecongiro** [comment]
@estebank Looks like #40683 splitted ` src/librustc_typeck/coherence/inherent.rs ` into two files and the hard error code disappeared when I rebased without paying attention. Sorry for confusion. I will create a PR shortly.

**Pull request merged**